File: app/src/main/java/agersant/polaris/api/remote/APIBase.java
package agersant.polaris.api.remote;

import android.net.Uri;

import com.google.android.exoplayer2.source.MediaSource;

import java.io.IOException;
import java.io.InputStream;

import agersant.polaris.CollectionItem;
import agersant.polaris.api.ItemsCallback;
import agersant.polaris.api.ThumbnailSize;
import okhttp3.Request;
import okhttp3.ResponseBody;

abstract class APIBase implements IRemoteAPI {

    private final DownloadQueue downloadQueue;
    final RequestQueue requestQueue;

    APIBase(DownloadQueue downloadQueue, RequestQueue requestQueue) {
        this.downloadQueue = downloadQueue;
        this.requestQueue = requestQueue;
    }

    abstract String getAudioURL(String path);

    abstract String getThumbnailURL(String path, ThumbnailSize size);

    public Uri getAudioUri(String path) {
        String url = getAudioURL(path);
        return Uri.parse(url);
    }

    public Uri getThumbnailUri(String path, ThumbnailSize size) {
        String url = getThumbnailURL(path, size);
        return Uri.parse(url);
    }

    public InputStream getThumbnail(String path, ThumbnailSize size) throws IOException {
        Request request = new Request.Builder().url(getThumbnailUri(path, size).toString()).build();
        ResponseBody body = requestQueue.requestSync(request);
        if (body != null) {
            return body.byteStream();
        } else {
            return null;
        }
    }

    abstract void getAlbums(String url, final ItemsCallback handlers);

    public void getRandomAlbums(ItemsCallback handlers) {
        String requestURL = ServerAPI.getAPIRootURL() + "/random/";
        getAlbums(requestURL, handlers);
    }

    public void getRecentAlbums(ItemsCallback handlers) {
        String requestURL = ServerAPI.getAPIRootURL() + "/recent/";
        getAlbums(requestURL, handlers);
    }

    public MediaSource getAudio(CollectionItem item) {
        return downloadQueue.getAudio(item);
    }

}

File: app/src/main/java/agersant/polaris/api/remote/APIBase.kt
package agersant.polaris.api.remote

import agersant.polaris.CollectionItem
import agersant.polaris.Directory
import agersant.polaris.Song
import agersant.polaris.api.ThumbnailSize
import agersant.polaris.api.remote.ServerAPI.Companion.apiRootURL
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import com.google.android.exoplayer2.source.MediaSource
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.utils.io.jvm.javaio.*

abstract class APIBase(
    private val downloadQueue: DownloadQueue,
    protected val client: HttpClient,
) : IRemoteAPI {

    abstract fun getAudioUrl(path: String): String

    abstract fun getThumbnailUrl(path: String, size: ThumbnailSize): String

    protected abstract suspend fun getAlbums(url: String): List<Directory>?

    override suspend fun getAudioUri(path: String): Uri? {
        val url = getAudioUrl(path)
        return Uri.parse(url)
    }

    override suspend fun getAudio(item: Song): MediaSource? {
        return downloadQueue.getAudio(item)
    }

    override suspend fun getThumbnail(path: String, size: ThumbnailSize): Bitmap? {
        return try {
            val response: HttpResponse = client.get(getThumbnailUrl(path, size))
            val stream = response.content.toInputStream()
            BitmapFactory.decodeStream(stream)
        } catch (e: Exception) {
            println("")
            null
        }
    }

    override suspend fun getRandomAlbums(): List<Directory>? {
        val url = "$apiRootURL/random/"
        return getAlbums(url)
    }

    override suspend fun getRecentAlbums(): List<Directory>? {
        val url = "$apiRootURL/recent/"
        return getAlbums(url)
    }
}
